service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -1651,4 +1700,34 @@ public final ConnectionLimitManager getConnectionLimitManager() {
     public final void setConnectionLimitManager(final ConnectionLimitManager connectionLimitManager) {
         this.connectionLimitManager = connectionLimitManager;
     }
+
+    private boolean wasAutoProvisioned(final JsonObject registrationInfo) {
+        if (registrationInfo != null) {
+            final JsonObject deviceDefaults = registrationInfo.getJsonObject(RequestResponseApiConstants.FIELD_PAYLOAD_DEFAULTS);
+            if (deviceDefaults != null) {
+                final String registrationStatus = deviceDefaults.getString(RegistrationConstants.REGISTRATION_STATUS.KEY);
+                if (RegistrationConstants.REGISTRATION_STATUS.NEW.name().equalsIgnoreCase(registrationStatus)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Adds auto-provisioning application properties to the given message, if present in the given context.
+     *
+     * @param msg The message for which the application properties shall be set.
+     * @param context The context containing the properties set in the processing of the auto-provisionoing.
+     */
+    protected void addAutoProvisioningProperties(final Message msg, final ExecutionContext context) {
+        if (context.get(MessageHelper.APP_PROPERTY_GATEWAY_ID, null) != null) {
+            MessageHelper.addProperty(msg, MessageHelper.APP_PROPERTY_GATEWAY_ID, context.get(MessageHelper.APP_PROPERTY_GATEWAY_ID));
+        }
+
+        if (context.get(RegistrationConstants.REGISTRATION_STATUS.KEY, null) != null) {
+            MessageHelper.addProperty(msg, RegistrationConstants.REGISTRATION_STATUS.KEY, context.get(RegistrationConstants.REGISTRATION_STATUS.KEY));
+        }
```

Comment:
so, in case of a telemetry message being used for auto-provisioning a device, the downstream application will receive two messages containing this property: the empty notification event and the telemetry message, right?

Reply:
Yes, that is correct. 

After a second thought: I guess setting the property in the telemetry message might be redundant since an application should always act upon an event with that property, since the event will be delivered at least once.

So I tend to not set the property for the telemetry message. WDYT?

Reply:
I am afraid that the whole mechanism as currently devised is not very robust. For example, if the connection between the adapter and the registry fails after the device has been provisioned but before the response message containing the registration info has reached the adapter, then the downstream application (Ditto) will not be notified about the newly provisioned device. However, subsequent requests from the gateway on behalf of the device will not trigger auto-provisioning again because the device now exists in the registry. Thus, Ditto will effectively never be notified about the new device.

Reply:
Yes, that is a potential edge case.

An implementation without such edge cases would involve keeping persistent state as far as I can see. We could keep a flag in the Device Registry Service implementation which indicates if the registration event was already sent downstream.

The protocol adapters then would need to check that flag as well and send an event as long as it is set to "pending delivery". When the event was successfully delivered to the AMQP network, the protocol adapters may set that flag to "delivered" with another call to the Device Registry Service.

However there may be still edge cases I guess, since the AMQP network "ack" is not an application-level "ack" - yet I think this should be sufficient.

Reply:
FMPOV all of this actually makes it more reasonable to let the registry emit the event that a device has been provisioned on-the-fly. We already follow that approach for auto-provisioning devices based on client certs, so this would actually be the more consistent approach anyway ...

Reply:
Maybe I missed something but I do not think that the device registry sends an event for auto-provisioning based on client certificates. It creates the device as well which is consistent with the implementation for gateway-based auto-provisioning.

From the perspective of the device registries' responsibilities it makes sense to put it there - however there is one thing which the device registry does not know: was it an event or a telemetry message? (i.e. do I need to send an event or not)

I found that rather odd to add a flag or the endpoint to the device registry's api since it doesn't have anything in common with device registration as far as I can see.

Or did you have a different approach in mind?

Reply:
> Maybe I missed something but I do not think that the device registry sends an event for auto-provisioning based on client certificates. It creates the device as well which is consistent with the implementation for gateway-based auto-provisioning.

Our example implementation (currently) doesn't, that's right. However, when we talked about how we wanted the client cert auto-provisioning to work, we agreed that it would be the responsibility of the registry implementation to notify other applications about the newly provisioned device. We left open how that should be done, but you can easily imagine that publishing an event would be a feasible option.

> From the perspective of the device registries' responsibilities it makes sense to put it there - however there is one thing which the device registry does not know: was it an event or a telemetry message? (i.e. do I need to send an event or not)

It doesn't need to because the registry would *always* publish an event indicating that the device has been provisioned while the protocol adapters would simply forward the telemetry message it received from the device *as is*.

> I found that rather odd to add a flag or the endpoint to the device registry's api since it doesn't have anything in common with device registration as far as I can see.

If the registry is responsible for emitting the event, there is no need to expose such a flag via its API.

Reply:
The implementation is now adapted so that it emits an empty event, so I guess this can be resolved?